providers/duo/v4.go

```diff
@@ -0,0 +1,230 @@
+package duo
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/tidwall/gjson"
+	"golang.org/x/net/html"
+)
+
+func executeAndParseJSON[T any](client *http.Client, req *http.Request) (dst T, err error) {
+	req.Header.Set("Accept", "application/json")
+	resp, err := client.Do(req)
+	if err != nil {
+		return
+	}
+
+	defer resp.Body.Close()
+	err = json.NewDecoder(resp.Body).Decode(&dst)
+	return
+}
+
+// DuoV4 contains functions for interacting with the Duo v4 Frame API.
+type DuoV4 struct {
+	BaseURL *url.URL
+	Client  *http.Client
+}
+
+func (d DuoV4) CheckStatusRequest(ctx context.Context, session Session) (*http.Request, error) {
+	vals := url.Values{
+		"txid": []string{session.Txid},
+		"sid":  []string{session.Sid},
+	}
+
+	next := d.BaseURL.ResolveReference(&url.URL{Path: "/frame/v4/status"})
+	req, err := http.NewRequestWithContext(ctx, "POST", next.String(), strings.NewReader(vals.Encode()))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	return req, nil
+}
+
+type CheckStatusResponse struct{}
+
+// CheckStatus checks the MFA status of the given txid/sid in Duo.
+//
+// The return type is not implemented because it is discarded by KeyConjurer.
+func (d DuoV4) CheckStatus(ctx context.Context, session Session) (CheckStatusResponse, error) {
+	req, _ := d.CheckStatusRequest(ctx, session)
+	return executeAndParseJSON[CheckStatusResponse](d.Client, req)
+}
+
+func (d DuoV4) WaitForPushAcknowledgement(ctx context.Context, session Session) error {
+	// TODO: Handle rejections - If we don't handle rejections, somewhere later in the application will fail with a cryptic error.
+	// This may not require two check statuses, if one was already called - but in normal functioning, this will always require at least two.
+	_, err := d.CheckStatus(ctx, session)
+	if err != nil {
+		return err
+	}
+
+	// The second one blocks until the push is approved.
+	// This second request will include cookies that we need to use in subsequent requests.
+	_, err = d.CheckStatus(ctx, session)
+	return err
+}
+
+type DuoV4APIResponse[T any] struct {
+	Stat     string
+	Response T
+}
+
+type Factor struct {
+	Key   string
+	Index string
+}
+
+type RegisteredFactors struct {
+	Phones []Factor
+}
+
+type Session struct {
+	Sid  string
+	Txid string
+	Xsrf string
+}
+
+// OpenFromResponse is like Open, but it consumes the given response rather than issuing one.
+//
+// This is useful when you know you have been redirected to the Duo session Auth page.
+func (d DuoV4) AuthFromResponse(ctx context.Context, resp *http.Response) (Session, error) {
+	defer resp.Body.Close()
```

Comment:
Not a bit deal but I'd prefer if the body were closed by the method that created it (even if it is a deferred call). I am not sure if Go has any best practice about it.

Reply:
It depends - if you're passing a http.Response to another function, you don't really know what that function is going to do with that response or whether you should be closing it.

However, the guidance - which I have only just learned after reading it - is that you should _always_ close the response body when using http.Client. A response body will _always_ be non-nil if the http.Do call was not an error, and you need to close it if the response.Body is ever not fully consumed otherwise you will get a resource leak.